### DISCUSSION TITLE
Installing IRAF on M1 MacBook Air

### DISCUSSION THREAD

**Discussion by Shaunak-PG2001** [iraf-community/iraf#178] [Q&A] [answered]
Hello!
I downloaded the updated distribution of IRAF community which supports ARM based M1 Macs. I followed the installation instructions correctly and at the end ran inbuilt tests using `./test/run_tests` command. However, it is throwing me an error that there is no `ecl.e` file in `bin.macos64`.  Please help me in resolving this issue. 

I had set the architecture properly using `make macos64`. So why is it that `ecl.e` is not in `bin.macos64` folder if `run_tests` expects it to be there?

[build.log](https://github.com/iraf-community/iraf/files/6668245/build.log)

**Comment by olebole** [comment]
The first error I find is (line 1061)
```
Fatal compiler error: Cannot execute xpp.e
```
which may have several causes:
 * there is no `xpp.e` executable in `unix/bin.macos64`
 * the symbolic link in `~/.iraf/bin/xpp` does not point to the proper executable
 * for some reason, another path is used here

Can you check that?

**Comment by Shaunak-PG2001** [comment, marked as answer]
1. The xpp.e executable is present in unix/bin.macos64
2. Yes, I guess the symbolic link `~/.iraf/bin/xpp` doesn't point to the proper executable because when I try to open it, I get the pop-up: The operation can’t be completed because the original item for “xpp” can’t be found.

**Reply by olebole** [reply]
Can you just show what comes out of `ls -l ~/.iraf/bin/xpp`? Maybe it points somewhere else?

**Reply by Shaunak-PG2001** [reply]
`lrwxr-xr-x  1 shaunakpadhyegurjar  staff  83 Jun 17 12:04 /Users/shaunakpadhyegurjar/.iraf/bin/xpp -> /Users/shaunakpadhyegurjar/Downloads/iraf-2.16.1-2021.06.14/unix/bin.macintel/xpp.e`

**Reply by olebole** [reply]
This shows that it is linked to the `unix/bin.macintel/` dir, as probably all the other ones. Do you have the environment variable `IRAFARCH` set, and what is the output of `./unix/hlib/irafarch.sh`?

**Reply by Shaunak-PG2001** [reply]
Yes, I had set `IRAFARCH` to `macos64`. Output of `./unix/hlib/irafarch.sh` is `macos64`

**Reply by olebole** [reply]
Can you also try `./unix/hlib/irafarch.sh -hsi`?

**Reply by Shaunak-PG2001** [reply]
Output of `./unix/hlib/irafarch.sh -hsi` is `macos64` too.

**Reply by olebole** [reply]
Then it is strange where the `bin.macintel` comes from here. When you did run `./install`, what was shown as Architecture and HSU arch?
```
========================================================================
=====================  Verifying System Settings  ======================
========================================================================

Hostname      = ...                OS version    = ....
Architecture  = macos64            HSI arch      = macos64             
Old iraf root = […]/iraf/iraf  New iraf root = […]/iraf/iraf
Old imdir     = /home/ole/.iraf/imdir  New imdir     = /home/ole/.iraf/imdir
Old cache     = /home/ole/.iraf/cache  New cache     = /home/ole/.iraf/cache

Local bin dir = /home/ole/.iraf/bin/
```
You can just re-run `./install` and stop at the question `Proceed with installation?  (yes):`.

**Reply by Shaunak-PG2001** [reply]
Well, I don't remember what was shown as architecture and HSU arch when I installed it before. So now I re-ran `./install` as you said and stopped at the question `Proceed with installation? (yes):`. Here's the output:
<img width="1080" alt="Screenshot 2021-06-17 at 2 02 45 PM" src="https://user-images.githubusercontent.com/63835398/122361614-0bbd2300-cf75-11eb-9873-35dd06d81aa4.png">

**Reply by olebole** [reply]
Then you can just re-run this and then check what is in `.iraf/bin`; whether it is now linked to `unix/bin.macos64/`.

**Reply by Shaunak-PG2001** [reply]
That means I need to enter 'yes' and then execute the commands `make macos64` and `make sysgen 2>&1 | tee build.log`, right?

**Reply by olebole** [reply]
To shorten up, I would recommend to check the content of `~/.iraf/bin/` before `make sysgen` and  make sure that they point to `bin.macos64` dirs.

**Reply by Shaunak-PG2001** [reply]
I checked, the aliases in `~/.iraf/bin/` pointed to correct files in `bin.macos64`. This time there was no error in build.log file but all the tests failed. 
<img width="704" alt="Screenshot 2021-06-17 at 2 38 02 PM" src="https://user-images.githubusercontent.com/63835398/122366954-a91a5600-cf79-11eb-9179-0894f1e6e784.png">

**Reply by olebole** [reply]
Cay you again attach the build log please?

**Reply by Shaunak-PG2001** [reply]
[build.log](https://github.com/iraf-community/iraf/files/6668961/build.log)

**Reply by olebole** [reply]
That seems OK.
Can you also attach the complete log of the run test and not just a screen shot? It seems that the tests are confused by some environment variables, like `iraf`, or `hbin` or so (maybe set from an old attempt to install IRAF). Can you check that, and maybe also attach the whole list of environment variables set?

**Reply by olebole** [reply]
One general comment: IRAF is really legacy code. I would not recommend to start anything new with IRAF. This will be a dead end. Instead, consider Astropy for analysis, which is already quite complete. Time is better invested in improving astropy than in trying to get IRAF running.

**Reply by Shaunak-PG2001** [reply]
Yes, indeed. I have wasted last few days in trying to install IRAF. Although, I require it for my internship, so I have to get it installed for now. Anyway, I have attached the tests and environment variables log file below.
[test_and_evnvariables.log](https://github.com/iraf-community/iraf/files/6669123/test_and_evnvariables.log)

**Reply by olebole** [reply]
That is some weird message from the Z-shell. You can get rid of it by doing
```shell
setopt +o nomatch
```
before running `./test/run_tests`.

**Reply by Shaunak-PG2001** [reply]
Nope. Still the test is failing.
[test2.log](https://github.com/iraf-community/iraf/files/6669283/test2.log)

**Reply by olebole** [reply]
Could you try to replace the exclamation mark `!` in https://github.com/iraf-community/iraf/blob/bcb80b314cbd8c4ee253abec1e2ca64645428579/unix/hlib/extpkg.cl#L26

with a double exclamation mark `!!`?

**Reply by Shaunak-PG2001** [reply]
Thanks! That made the installation pass all tests. Could you do just one more favour for me? Could you please tell me how to start IRAF? The method given in installation script (see the log file below) doesn't work (`source ~/.login` followed by `cl`). Also, typing `xgterm` in terminal says `bad CPU type in executable: xgterm`. Please tell me the proper way of starting IRAF. Thanks a lot! 
[start_iraf.log](https://github.com/iraf-community/iraf/files/6669617/start_iraf.log)

**Reply by olebole** [reply]
The first response in your file is
```
shaunakpadhyegurjar@Shaunaks-MacBook-Air ~ % source ~/.login
source: no such file or directory: /Users/shaunakpadhyegurjar/.login
```
which lets me assum that you don't have a `.login` file. Probably you need to source `~/.bashrc`.

And the next one
```
shaunakpadhyegurjar@Shaunaks-MacBook-Air ~ % cl
ERROR:  No /opt/iraf/iraf//bin.ipad/vocl.e binary found.
```
shows that you have some old IRAF paths left over from somewhere else: somehow iraf is set to `/opt/iraf/iraf/`, and the architecture is set to `ipad`, which definitely does not come from the just-installed IRAF. It probably comes from an old installation, where the [Mac architecture defaults to `ipad` if the machine is not an x86](https://github.com/iraf-community/iraf/blob/c2e107fc404aa26d135741357882d19744512815/unix/hlib/irafarch.sh#L127-L150).

I am afraid this is something you need to clean up yourself on your machine.

For xgterm: where did you get the executable from? The old binary packages don't work on an M1 since they are 32 bit only. You need to compile yourself, see https://iraf-community.github.io/x11iraf.

**Reply by Shaunak-PG2001** [reply]
Hello. Just doing `source ~/.bashrc` followed by `cl` did the job. I am very greatful to you. Thank you very much for staying with me upto this point. Thank you, sir!
